File: python/src/convalesce_emit/spool.py
"""
Local disk for batches that could not be delivered.

A batch the receiver could not take -- it was down, the key was wrong, the
network was out -- is written here instead of being dropped, and sent again
the next time a send succeeds, from this process or any other on the host. A
batch the receiver understood and refused (400, 413, 422) goes to a separate
`rejected` directory: sending it again would be refused again, but it is kept
for whoever needs to see what was in it.

Every file is one gzip-compressed request body, exactly as it would have gone
on the wire, so replaying one is a single POST.

Import as:

import convalesce_emit.spool as cespool
"""

import logging
import os
import time
import uuid
from typing import List, Optional

_LOG = logging.getLogger(__name__)

PENDING = "pending"
REJECTED = "rejected"
_SUFFIX = ".json.gz"
_CLAIMED = ".sending"
# A claim older than this belongs to a process that died mid-send; the batch
# goes back in the queue rather than sitting claimed forever.
_STALE_CLAIM_SECONDS = 600


class Spool:
    """
    Undelivered batches on local disk, oldest first.

    :param directory: where the `pending` and `rejected` directories live
    :param max_bytes: total size `pending` may reach; past it a new batch is
        refused with an error in the log, because filling a customer's disk
        would do more harm than the batch is worth
    """

    def __init__(self, directory: str, max_bytes: int) -> None:
        self.directory = directory
        self.max_bytes = max_bytes

    def save(self, body: bytes, kind: str = PENDING) -> Optional[str]:
        """
        Write one compressed request body.

        :param body: the gzip-compressed body
        :param kind: `PENDING` to send again, `REJECTED` to keep only
        :return: the file written, or None when it could not be
        """
        folder = os.path.join(self.directory, kind)
        try:
            os.makedirs(folder, exist_ok=True)
            if (
                kind == PENDING
                and self._size(folder) + len(body) > self.max_bytes
            ):
                _LOG.error(
                    "convalesce: spool %s is past %d bytes; a batch of %d "
                    "bytes could not be kept",
                    folder,
                    self.max_bytes,
                    len(body),
                )
                return None
            name = f"{time.time_ns():020d}-{uuid.uuid4().hex}{_SUFFIX}"
            path = os.path.join(folder, name)
            partial = path + ".partial"
            with open(partial, "wb") as out:
                out.write(body)
            # Renamed into place so a reader never sees half a file.
            os.replace(partial, path)
            return path
        except OSError as exc:
            _LOG.error(
                "convalesce: could not spool a batch to %s: %s", folder, exc
            )
            return None

    def pending(self) -> List[str]:
        """
        Every batch waiting to be sent again, oldest first.

        Claims left behind by a process that died are released first.

        :return: file paths
        """
        folder = os.path.join(self.directory, PENDING)
        try:
            names = os.listdir(folder)
        except OSError:
            return []
        now = time.time()
        out: List[str] = []
        for name in names:
            path = os.path.join(folder, name)
            if name.endswith(_SUFFIX):
                out.append(path)
            elif _CLAIMED in name:
                try:
                    if now - os.path.getmtime(path) > _STALE_CLAIM_SECONDS:
                        original = path[: path.index(_CLAIMED)]
                        os.replace(path, original)
                        out.append(original)
                except (OSError, ValueError):
                    continue
        return sorted(out)

    def claim(self, path: str) -> Optional[str]:
        """
        Take one batch so no other process sends it at the same time.

        :param path: a path from `pending()`
        :return: the claimed path, or None when another process got it first
        """
        claimed = f"{path}{_CLAIMED}.{os.getpid()}"
        try:
            os.replace(path, claimed)
            os.utime(claimed)
        except OSError:
            return None
        return claimed

    def release(self, claimed: str) -> None:
        """
        Put a claimed batch back, to be tried again later.

        :param claimed: a path from `claim()`
        """
        try:
            os.replace(claimed, claimed[: claimed.index(_CLAIMED)])
        except (OSError, ValueError) as exc:
            _LOG.error(
                "convalesce: could not return %s to the spool: %s", claimed, exc
            )

    def reject(self, claimed: str) -> None:
        """
        Move a claimed batch the receiver refused to `rejected`.

        :param claimed: a path from `claim()`
        """
        folder = os.path.join(self.directory, REJECTED)
        try:
            os.makedirs(folder, exist_ok=True)
            name = os.path.basename(claimed[: claimed.index(_CLAIMED)])
            os.replace(claimed, os.path.join(folder, name))
        except (OSError, ValueError) as exc:
            _LOG.error(
                "convalesce: could not keep refused batch %s: %s", claimed, exc
            )

    @staticmethod
    def read(path: str) -> bytes:
        """
        The compressed body in one file.

        :param path: a claimed path
        :return: its bytes
        """
        with open(path, "rb") as source:
            return source.read()

    @staticmethod
    def done(claimed: str) -> None:
        """
        Remove a batch the receiver accepted.

        :param claimed: a path from `claim()`
        """
        try:
            os.remove(claimed)
        except OSError:
            pass

    @staticmethod
    def _size(folder: str) -> int:
        total = 0
        for name in os.listdir(folder):
            try:
                total += os.path.getsize(os.path.join(folder, name))
            except OSError:
                continue
        return total

File: java/emit-core/src/main/java/io/convalesce/emit/Spool.java
package io.convalesce.emit;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;
import java.util.logging.Logger;

/**
 * Local disk for batches that could not be delivered.
 *
 * <p>The same layout the Python client uses, so one host running both shares one queue: {@code
 * pending/} holds batches to send again after the next send that succeeds, {@code rejected/} holds
 * batches the receiver read and refused, kept rather than dropped. Each file is one gzip-compressed
 * request body, exactly as it would have gone on the wire.
 */
final class Spool {

  static final String PENDING = "pending";
  static final String REJECTED = "rejected";

  private static final Logger LOG = Logger.getLogger(Spool.class.getName());
  private static final String SUFFIX = ".json.gz";
  private static final String CLAIMED = ".sending";
  // A claim older than this belongs to a process that died mid-send.
  private static final long STALE_CLAIM_MS = 600_000L;

  private final File directory;
  private final long maxBytes;

  Spool(String directory, long maxBytes) {
    this.directory = new File(directory);
    this.maxBytes = maxBytes;
  }

  /**
   * Writes one compressed request body.
   *
   * @return the file written, or null when it could not be kept
   */
  File save(byte[] body, String kind) {
    File folder = new File(directory, kind);
    try {
      if (!folder.isDirectory() && !folder.mkdirs() && !folder.isDirectory()) {
        throw new IOException("cannot create " + folder);
      }
      if (PENDING.equals(kind) && size(folder) + body.length > maxBytes) {
        LOG.severe(
            "convalesce: spool "
                + folder
                + " is past "
                + maxBytes
                + " bytes; a batch of "
                + body.length
                + " bytes could not be kept");
        return null;
      }
      String name =
          String.format("%020d", System.currentTimeMillis() * 1_000_000L)
              + "-"
              + UUID.randomUUID().toString().replace("-", "")
              + SUFFIX;
      File path = new File(folder, name);
      File partial = new File(folder, name + ".partial");
      FileOutputStream out = new FileOutputStream(partial);
      try {
        out.write(body);
      } finally {
        out.close();
      }
      // Moved into place so a reader never sees half a file.
      Files.move(partial.toPath(), path.toPath(), StandardCopyOption.ATOMIC_MOVE);
      return path;
    } catch (IOException e) {
      LOG.severe("convalesce: could not spool a batch to " + folder + ": " + e.getMessage());
      return null;
    }
  }

  /** Every batch waiting to be sent again, oldest first, after releasing stale claims. */
  List<File> pending() {
    File[] files = new File(directory, PENDING).listFiles();
    List<File> out = new ArrayList<File>();
    if (files == null) {
      return out;
    }
    long now = System.currentTimeMillis();
    for (File file : files) {
      String name = file.getName();
      if (name.endsWith(SUFFIX)) {
        out.add(file);
      } else if (name.contains(CLAIMED) && now - file.lastModified() > STALE_CLAIM_MS) {
        File original = new File(file.getParentFile(), name.substring(0, name.indexOf(CLAIMED)));
        if (file.renameTo(original)) {
          out.add(original);
        }
      }
    }
    Collections.sort(out);
    return out;
  }

  /**
   * Takes one batch so no other process sends it at the same time.
   *
   * @return the claimed file, or null when another process got it first
   */
  File claim(File path) {
    File claimed = new File(path.getParentFile(), path.getName() + CLAIMED + "." + pid());
    if (!path.renameTo(claimed)) {
      return null;
    }
    claimed.setLastModified(System.currentTimeMillis());
    return claimed;
  }

  /** Puts a claimed batch back, to be tried again later. */
  void release(File claimed) {
    String name = claimed.getName();
    File original = new File(claimed.getParentFile(), name.substring(0, name.indexOf(CLAIMED)));
    if (!claimed.renameTo(original)) {
      LOG.severe("convalesce: could not return " + claimed + " to the spool");
    }
  }

  /** Moves a claimed batch the receiver refused to {@code rejected/}. */
  void reject(File claimed) {
    File folder = new File(directory, REJECTED);
    folder.mkdirs();
    String name = claimed.getName();
    if (!claimed.renameTo(new File(folder, name.substring(0, name.indexOf(CLAIMED))))) {
      LOG.severe("convalesce: could not keep refused batch " + claimed);
    }
  }

  static byte[] read(File claimed) throws IOException {
    return Files.readAllBytes(claimed.toPath());
  }

  static void done(File claimed) {
    if (!claimed.delete()) {
      LOG.warning("convalesce: could not remove delivered batch " + claimed);
    }
  }

  private static long size(File folder) {
    long total = 0;
    File[] files = folder.listFiles();
    if (files != null) {
      for (File file : files) {
        total += file.length();
      }
    }
    return total;
  }

  private static String pid() {
    // "pid@host" on every JVM this targets; Java 8 has no ProcessHandle.
    String name = java.lang.management.ManagementFactory.getRuntimeMXBean().getName();
    int at = name.indexOf('@');
    return at > 0 ? name.substring(0, at) : name;
  }
}
